rfcs/0023-adding-annotation-field-to-tir.allocate.md

```diff
@@ -0,0 +1,201 @@
+        __attribute__((section( "FLASH" ), aligned( 16 )))  const uint8_t my_model_params_2[TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE] = <param_2_data>;
+        __attribute__((section( "DTCM" )  static uint8_t workspace_buffer_1[TVM_MY_MODEL_DTCM_WORKSPACE_BUFFER_SIZE];
+        __attribute__((section( "SRAM" ), aligned( 16 )))  static uint8_t workspace_buffer_2[TVM_MY_MODEL_SRAM_WORKSPACE_BUFFER_SIZE];
+
+    int main(...) {
+         ...
+         TVMContext context;
+         TVMInputs_my_model_1 inputs = {input};
+         TVMOutputs_my_model_1 outputs = {output};
+         TVMWorkspaces_my_model workspaces = {
+             .sram = &workspace_buffer_1,
+             .dtcm = &workspace_buffer_2,
+         };
+         TVMParameters_my_model parameters = {
+             .flash = &my_model_params_1,
+             .itcm = &my_model_params_2
+         };
+         TVMSetWorkspaces(&context, &workspaces);
+         TVMSetParameters(&context, parameters);
+         TVMExecute(&my_model, &inputs, &outputs, &context);
+    }
+```
+
+Therefore, we'd need a way to represent the association of each of these memories, that the user will pin the buffers to, closer to allocate nodes in TIR.
+
+At the IR, we ll need to associate each allocate node with one (or more) memories that it can end up, because the scheduling might be satisfied with placing buffers in any of the memories in a given set of memories. Therefore, the scheduler might want the memory planners to decide which memory to use based on finding the allocation that fit.
+
+ There are broadly two requirements here :
+
+P1 : Indicate candidate memories (a.k.a. Pools) that each allocate be associated with
+
+P2 : Indicate the final memory the allocate will be pinned on
+
+
+To serve P2, we propose to use the existing tag of the storage_scope with 'global-<memory_name>'.
+
+```
+struct StorageScope {
+  /*! \brief The rank of the storage */
+  StorageRank rank{StorageRank::kGlobal};
+  /*! \brief tag for special purpose memory. */
+  std::string tag;
+```
+
+To serve P1, this RFC introduces the addition of the annotations field
+
+# 3. Guide-level explanation
+
+This is not particularly a user-facing feature.
+
+
+ # 4. Reference-level explanation
+
+
+To serve P1, we propose to use :
+
+```
+class AllocateNode : public StmtNode {
+ public:
+  /*! \brief The buffer variable. */
+  Var buffer_var;
+  /*! \brief The type of the buffer. */
+  DataType dtype;
+  /*! \brief The extents of the buffer. */
+  Array<PrimExpr> extents;
+  /*! \brief Only allocate buffer when condition is satisfied. */
+  PrimExpr condition;
+  /*! \brief The body to be executed. */
+  Stmt body;
+  /*! \brief If the allocate is scoped global, this field indicates
+   *  which external memories it could be pinned to as a comma seperated
+   *  string.
+   */
++ Map<String, Objectref> annotations;
```

Comment:
I'm not sure I love having both the attr mechanism and this arbitrary map<str, value>. Can you say why we should use ObjectRef as the value here?

Reply:
I think we are being consistent with other TIR nodes and this (and others) use ObjectRef as a general way to pass compilation hints from IR point of view while the passes could specialize that. I personally agreed with @tqchen on the basis that this avenue will create the lesser changes in future for the IR, rather than doing a specialized addition of PoolInfo here.

The question of having both ways of annotations, seems to be out of scope for this RFC as we are not introducing another way here.

(if you follow @tqchen's citation of ForNode)

Reply:
Ah okay nevermind, I traced through more things and I agree with ObjectRef as a value.